iotdb/src/main/java/org/apache/iotdb/db/utils/datastructure/ListPublicBAOS.java

```diff
@@ -0,0 +1,193 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.db.utils.datastructure;
+
+import static org.apache.iotdb.db.utils.datastructure.ListPublicBAOSPool.ARRAY_SIZE;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.iotdb.tsfile.utils.PublicBAOS;
+
+
+/**
+ * We reimplement {@linkplain PublicBAOS}, replacing the underlying {@linkplain
+ * java.io.ByteArrayOutputStream} with a {@code List } of {@code byte[]}.
+ * <p>
+ * For efficient and controllable GC, all {@code byte[]} are allocated from {@linkplain
+ * ListPublicBAOSPool} and should be put back after {@code close}.
+ * <p>
+ * Referring to {@linkplain TVList} and {@linkplain org.apache.iotdb.db.rescon.PrimitiveArrayPool PrimitiveArrayPool}.
+ * <p>
+ * So far, this class is only used in {@linkplain org.apache.iotdb.db.engine.memtable.MemTableFlushTask
+ * MemTableFlushTask}.
+ *
+ * @author Pengze Lv, kangrong
+ */
+public class ListPublicBAOS extends PublicBAOS {
+
+  /**
+   * data buffer
+   */
+  private List<byte[]> values;
+
+  /**
+   * The number of valid bytes in the buffer.
+   */
+  private int dataSize;
+
+  /**
+   * capacity of the buffer.
+   */
+  private int capacity;
+
+  public ListPublicBAOS() {
+    super();
+
+    values = new ArrayList<>();
+    dataSize = 0;
+    capacity = 0;
+  }
+
+  public ListPublicBAOS(int size) {
+    super();
+    if (size < 0) {
+      throw new IllegalArgumentException("Negative initial size: "
+              + size);
+    }
+
+    values = ListPublicBAOSPool.getInstance().getByteLists(size);
+    dataSize = 0;
+    capacity = size;
+  }
+
+  @Override
+  public synchronized void write(byte[] b, int off, int len) {
```

Comment:
is synchronized needed?

Reply:
> is synchronized needed?

It's consistent with its parent class `ByteArrayOutputStream`.